File: apps/mobile-flutter/test/shared/logic/calorie_readout_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:kallo_mobile/models/nutrition/nutrition_enums.dart';
import 'package:kallo_mobile/shared/logic/calorie_readout.dart';

const double _target = 2000;

CalorieReadout _readout(double logged, MacroGoal? goal) =>
    calorieReadout(logged: logged, target: _target, goal: goal);

void main() {
  test('leads a cutter with what is left to spend', () {
    final readout = _readout(741, MacroGoal.cutting);

    expect(readout.headline, 1259);
    expect(readout.framing, CalorieFraming.remaining);
    expect(readout.left, 1259);
    expect(readout.over, isNull);
  });

  test('leads every other goal with what has been logged', () {
    for (final goal in [MacroGoal.bulking, MacroGoal.maintaining, null]) {
      final readout = _readout(741, goal);

      expect(readout.headline, 741, reason: '$goal');
      expect(readout.framing, CalorieFraming.logged, reason: '$goal');
      expect(readout.left, 1259, reason: '$goal');
    }
  });

  test('never hands a cutter a negative headline', () {
    final readout = _readout(2341, MacroGoal.cutting);

    expect(readout.headline, 0);
    // The deficit is spent, but the overshoot is not hidden.
    expect(readout.over, 341);
  });

  test('names the overshoot for a goal that counts up', () {
    final readout = _readout(2341, MacroGoal.bulking);

    expect(readout.headline, 2341);
    expect(readout.left, 0);
    expect(readout.over, 341);
  });
}

File: lib/domain/nutrition/__tests__/calorie-readout.test.ts
import { describe, expect, it } from 'vitest';
import { calorieReadout } from '@/lib/domain/nutrition/calorie-readout';
import type { Goal } from '@/lib/domain/onboarding/types';

const TARGET = 2000;

describe('calorieReadout', () => {
  it('leads a cutter with what is left to spend', () => {
    expect(calorieReadout(741, TARGET, 'cutting')).toEqual({
      headline: 1259,
      framing: 'remaining',
      left: 1259,
      over: null,
    });
  });

  it.each<Goal | null>([
    'bulking',
    'maintaining',
    null,
  ])('leads %s with what has been logged', (goal) => {
    expect(calorieReadout(741, TARGET, goal)).toMatchObject({
      headline: 741,
      framing: 'logged',
      left: 1259,
    });
  });

  it('never hands a cutter a negative headline', () => {
    const readout = calorieReadout(2341, TARGET, 'cutting');

    expect(readout.headline).toBe(0);
    // The deficit is spent, but the overshoot is not hidden.
    expect(readout.over).toBe(341);
  });

  it('names the overshoot for a goal that counts up', () => {
    expect(calorieReadout(2341, TARGET, 'bulking')).toMatchObject({
      headline: 2341,
      left: 0,
      over: 341,
    });
  });

  it('rounds before it decides, so the two figures always add up', () => {
    const readout = calorieReadout(740.6, TARGET, 'cutting');

    expect(readout.headline).toBe(1259);
    expect(readout.left).toBe(1259);
  });
});
